System/Posix/IO.hsc

```diff
@@ -111,3 +118,25 @@ createFileAt :: Maybe Fd -- ^ Optional directory file descriptor
              -> IO Fd
 createFileAt fdMay name mode
   = openFdAt fdMay name WriteOnly defaultFileFlags{ trunc=True, creat=(Just mode) }
+
+-- | Read data from an 'Fd' and convert it to a 'String' using the locale encoding.
+-- Throws an exception if this is an invalid descriptor, or EOF has been
+-- reached.
+fdRead :: Fd
```

Comment:
It may be worth noting that unless the input is known to be ASCII data, or an 8-bit locale is in use, reading `n` bytes and expecting a `String` is rather dubious.  One can read and decode *lines* (in an UTF-8 locale) because UTF-8 is self-synchronising, and LF or CRLF is never in the middle of a multi-byte sequence.  Otherwise, reading "n-bytes" from a file into a string is a recipe for failure.  Use of this function should be discouraged.

Reply:
I'm not sure I follow.

This issues seems to be present everywhere, including in the handling of filepaths. It's broken: https://gist.github.com/hasufell/c600d318bdbe010a7841cc351c835f92

Anything that uses `getLocaleEncoding`, `getFileSystemEncoding` or `getForeignEncoding` to convert `CString` to `String` potentially runs into bugs, no?

Following that, we'd need to deprecate half of base and most of the unix String based API.

Reply:
> Anything that uses `getLocaleEncoding`, `getFileSystemEncoding` or `getForeignEncoding` to convert `CString` to `String` potentially runs into bugs, no?

No, the issue here is that we're not reading a *complete* string, we're reading `n bytes` from a file.  These `n bytes` may well end in the middle of a multi-byte grapheme.

Converting *entire* file names or similar octet data (rather than `n byte` fragments of strings) to Strings is much safer under reasonable assumptions.

Reply:
> These n bytes may well end in the middle of a multi-byte grapheme.

Right, got it. And when you append the strings, you get garbage.

Reply:
> > These n bytes may well end in the middle of a multi-byte grapheme.
> 
> Right, got it. And when you append the strings, you get garbage.

Yes, the description of the action says it all: *read n bytes*.  That's a `ByteString` operation, not a `String` operation.  `String`s don't consist of bytes.